File: frontend/src/domain/entities/Node.ts
import { z } from "zod";

const nodeCode = z.string().min(1).max(32);

const BranchesSchema = z.object({
  yes: nodeCode,
  no: nodeCode,
});

const QuestionNodeSchema = z.object({
  id: nodeCode,
  type: z.literal("question"),
  text: z.string().min(1).max(1000),
  branches: BranchesSchema,
});

const LeafNodeSchema = z.object({
  id: nodeCode,
  type: z.literal("leaf"),
  outcome: z.enum(["PASS", "FAIL", "NOT_APPLICABLE"]),
  text: z.string().max(1000).optional(),
});

export const NodeSchema = z.discriminatedUnion("type", [QuestionNodeSchema, LeafNodeSchema]);

export type NodeRaw = z.infer<typeof NodeSchema>;
export type Branches = z.infer<typeof BranchesSchema>;
export type Outcome = "PASS" | "FAIL" | "NOT_APPLICABLE";

// Contratto astratto comune: QuestionNode e LeafNode lo implementano ciascuna
// in modo indipendente (nessuna gerarchia di classi, nessuno stato condiviso).
export interface NodeContract {
  readonly id: string;
  readonly type: "question" | "leaf";
  next(answer: boolean): string;
  verdict(): Outcome | null;
}

export class QuestionNode implements NodeContract {
  readonly #id: string;
  readonly #text: string;
  readonly #branches: Branches;

  constructor(id: string, text: string, branches: Branches) {
    this.#id = id;
    this.#text = text;
    this.#branches = branches;
  }

  get id(): string {
    return this.#id;
  }

  get type(): "question" {
    return "question";
  }

  get text(): string {
    return this.#text;
  }

  get branches(): Branches {
    return this.#branches;
  }

  next(answer: boolean): string {
    return answer ? this.#branches.yes : this.#branches.no;
  }

  verdict(): Outcome | null {
    return null;
  }

  toJSON() {
    return { id: this.id, type: this.type, text: this.text, branches: this.branches };
  }
}

export class LeafNode implements NodeContract {
  readonly #id: string;
  readonly #outcome: Outcome;
  readonly #text: string | undefined;

  constructor(id: string, outcome: Outcome, text?: string) {
    this.#id = id;
    this.#outcome = outcome;
    this.#text = text;
  }

  get id(): string {
    return this.#id;
  }

  get type(): "leaf" {
    return "leaf";
  }

  get outcome(): Outcome {
    return this.#outcome;
  }

  get text(): string | undefined {
    return this.#text;
  }

  next(): string {
    throw new TypeError("un nodo foglia non ha un nodo successivo");
  }

  verdict(): Outcome | null {
    return this.#outcome;
  }

  toJSON() {
    return { id: this.id, type: this.type, outcome: this.outcome, text: this.text };
  }
}

// Unione usata da tutti i chiamanti: consente il narrowing su `.type`
// (node.type === "leaf" => node ha .outcome), che una classe base astratta
// con `extends` non permetterebbe.
export type Node = QuestionNode | LeafNode;

export function createNode(raw: NodeRaw): Node {
  return raw.type === "question"
    ? new QuestionNode(raw.id, raw.text, raw.branches)
    : new LeafNode(raw.id, raw.outcome, raw.text);
}

File: backend/src/domain/node.py
from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from enum import Enum


class NodeOutcome(str, Enum):
    PASS = "PASS"
    FAIL = "FAIL"
    NOT_APPLICABLE = "NOT_APPLICABLE"

    @classmethod
    def from_string(cls, value: str) -> NodeOutcome:
        for member in cls:
            if member.value == value:
                return member
        raise ValueError(f"Unknown node outcome: {value}")


@dataclass(frozen=True)
class Branches:
    yes: str
    no: str


class Node(ABC):
    @property
    @abstractmethod
    def id(self) -> str: ...

    @abstractmethod
    def next(self, answer: bool) -> str: ...

    @abstractmethod
    def verdict(self) -> NodeOutcome | None: ...


@dataclass(frozen=True, init=False)
class QuestionNode(Node):
    _id: str
    _text: str
    _branches: Branches

    def __init__(self, id: str, text: str, branches: Branches, type: str = "question") -> None:
        object.__setattr__(self, "_id", id)
        object.__setattr__(self, "_text", text)
        object.__setattr__(self, "_branches", branches)

    @property
    def id(self) -> str:
        return self._id

    @property
    def type(self) -> str:
        return "question"

    @property
    def text(self) -> str:
        return self._text

    @property
    def branches(self) -> Branches:
        return self._branches

    def next(self, answer: bool) -> str:
        return self._branches.yes if answer else self._branches.no

    def verdict(self) -> NodeOutcome | None:
        return None


@dataclass(frozen=True, init=False)
class LeafNode(Node):
    _id: str
    _outcome: NodeOutcome
    _text: str | None

    def __init__(
        self, id: str, outcome: NodeOutcome | str, text: str | None = None, type: str = "leaf"
    ) -> None:
        object.__setattr__(self, "_id", id)
        object.__setattr__(self, "_outcome", NodeOutcome.from_string(outcome))
        object.__setattr__(self, "_text", text)

    @property
    def id(self) -> str:
        return self._id

    @property
    def type(self) -> str:
        return "leaf"

    @property
    def outcome(self) -> NodeOutcome:
        return self._outcome

    @property
    def text(self) -> str | None:
        return self._text

    def next(self, answer: bool) -> str:
        raise TypeError("un nodo foglia non ha un nodo successivo")

    def verdict(self) -> NodeOutcome | None:
        return self._outcome
